storage/statedb/database.go

```diff
@@ -541,6 +563,9 @@ func (db *Database) reference(child common.Hash, parent common.Hash) {
 
 // Dereference removes an existing reference from a root node.
 func (db *Database) Dereference(root common.Hash) {
+	db.gcLock.Lock()
```

Comment:
hmm.. I'm not 100% sure, but it seems that `bd.lock` is for `db.nodes`.

Should we have a different lock for gc?

Reply:
`database.lock` is for reading/writing only `database.nodes`.
But `gcLock` is for preventing to gc trie node in `database.nodes` without writing to persistent database.

In `StateAtWithGCLock`, with `RGCLock`, it returns the stateDB.
During the process, `DoesExistCachedNode` check if the state root is available with `database.lock`. 

After the function returns, the root can be removed by `Cap` or ` Commit`.
But in this cases, the root already was written on persistent database during `Cap`, `Commit`.

So, the stateDB is valid.